pkg/apis/coordination/validation/validation.go

```diff
@@ -48,5 +57,118 @@ func ValidateLeaseSpec(spec *coordination.LeaseSpec, fldPath *field.Path) field.
+	}
+	if spec.PreferredHolder != nil && spec.Strategy == nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("preferredHolder"), "may only be specified if `strategy` is defined"))
+	}
+	// spec.RenewTime is a MicroTime and doesn't need further validation
+	return allErrs
+}
+
+// ValidateLeaseCandidate validates a LeaseCandidate.
+func ValidateLeaseCandidate(lease *coordination.LeaseCandidate) field.ErrorList {
+	allErrs := validation.ValidateObjectMeta(&lease.ObjectMeta, true, ValidLeaseCandidateName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateLeaseCandidateSpec(&lease.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+func ValidLeaseCandidateName(name string, prefix bool) []string {
+	// prefix is already handled by IsConfigMapKey, a trailing - is permitted.
+	return utilvalidation.IsConfigMapKey(name)
+}
+
+func ValidateLeaseCandidateSpecUpdate(leaseCandidateSpec, oldLeaseCandidateSpec *coordination.LeaseCandidateSpec) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(leaseCandidateSpec.LeaseName, oldLeaseCandidateSpec.LeaseName, field.NewPath("leaseName"))...)
+	return allErrs
+}
+
+// ValidateLeaseCandidateUpdate validates an update of LeaseCandidate object.
+func ValidateLeaseCandidateUpdate(leaseCandidate, oldLeaseCandidate *coordination.LeaseCandidate) field.ErrorList {
+	allErrs := validation.ValidateObjectMetaUpdate(&leaseCandidate.ObjectMeta, &oldLeaseCandidate.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateLeaseCandidateSpec(&leaseCandidate.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateLeaseCandidateSpecUpdate(&leaseCandidate.Spec, &oldLeaseCandidate.Spec)...)
+	return allErrs
+}
+
+// ValidateLeaseCandidateSpec validates spec of LeaseCandidate.
+func ValidateLeaseCandidateSpec(spec *coordination.LeaseCandidateSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(spec.LeaseName) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("leaseName"), ""))
+	}
+	ev := semver.Version{}
+	if spec.EmulationVersion != "" {
+		var err error
+		ev, err = semver.Parse(spec.EmulationVersion)
+		if err != nil {
+			fld := fldPath.Child("emulationVersion")
+			allErrs = append(allErrs, field.Invalid(fld, spec.EmulationVersion, "must be a valid semantic version"))
+		}
+	}
+	bv := semver.Version{}
+	if spec.BinaryVersion != "" {
+		var err error
+		bv, err = semver.Parse(spec.BinaryVersion)
+		if err != nil {
+			fld := fldPath.Child("binaryVersion")
+			allErrs = append(allErrs, field.Invalid(fld, spec.BinaryVersion, "must be a valid semantic version"))
+		}
+	}
+	if !bv.EQ(ev) && bv.LT(ev) {
+		fld := fldPath.Child("binaryVersion")
+		allErrs = append(allErrs, field.Invalid(fld, spec.BinaryVersion, "must be greater than or equal to `emulationVersion`"))
+	}
+
+	strategiesMap := make(map[coordination.CoordinatedLeaseStrategy]bool)
```

Comment:
not a good variable name. What does it say about a stragegy? Is it `strategySeen` or just `seen`?